pkg/apis/core/v1alpha1/resource.go

```diff
@@ -40,6 +40,8 @@ const (
 	ResourceCredentialsSecretClientKeyKey = "clientKey"
 	// ResourceCredentialsTokenKey is the key inside a connection secret for the bearer token value
 	ResourceCredentialsTokenKey = "token"
+	// ResourceBucketNameKey is the key inside a connection secret for a bucket that identifies the name of the bucket
+	ResourceBucketNameKey = "bucketName"
```

Comment:
Aspirationally, connection secrets should be "generic" and not tied to any specific resource.
`ResourceBucketNameKey` - ties it explicitly to Bucket types.
Would you consider to use `Endpoint` instead as we do for GCP?

Reply:
https://github.com/crossplaneio/crossplane/blob/master/pkg/controller/gcp/storage/bucket_operations.go#L142

Reply:
Personally I think it makes the code a bit harder to grasp if we start down this path, and although google only has one endpoint for a bucket, s3's endpoints are region specific and we generate a real endpoint that is unique to that region. Would you like it moved to the bucket class and renamed BucketResourceNameKey or maybe call is ResourceNameKey? Other?

Reply:
I agree that the `Endpoint` approach is not perfect, but it is "less imperfect" than adding additional fields which probably won't be utilized in any other resources. Granted that the `connection secret` itself is not perfect and requires additional thought. 

GCP resource itself "piggy-backs" on the `Endpoint` field to save the bucket name, which is *NOT* an endpoint, since GCP bucket can be addressed in different ways also (`gds://` ` or `s3://` etc), it is the responsibility of the consuming application to construct correct "connection string" whichever it may be. 

With that said, I would prefer for us to follow the same paradigm for both GCP and AWS. If S3 consuming application needs to construct the "real" endpoint, it should have all the necessary ingredients in the connection secret. Thus, in this specific context: "ResourceEndpoint" should be semantically identical to what is proposed as "ResourceBucketName"

> s3's endpoints are region specific
I presume we were not intended to save "region" value into "ResourceBucketNameKey"

Reply:
I created https://github.com/crossplaneio/crossplane/issues/473 to keep track of this issue.

Reply:
I'm fairly sure that s3 bucket needs to have four fields to make a connection: user, pass, bucket, region(should fix in this PR - not endpoint). 

S3 proposed mapping:
user->user
key->key
endpoint->region
bucket->bucket

I think you're asking that if we add bucket, you would like to see gcp fixed as well to something like this:
user->interop user
key->interop key
token->gcs token
bucket->bucket

We should just be cautious that the gcp update will cause a ripple on gcp demos and gitlab-controller that we'll need to make sure to fix as well. Or take as a separate work item for next week.

Reply:
On another note if we were using minio, endpoint would be used as an endpoint

Reply:
What should I change in this PR? The build is passing, just need to handle this issue.

Reply:
@rbwsam I am not :100: comfortable on this specific change, i.e. introducing new resource-specific field. I don't fully understand @lukeweber objections in using `endpoint` field, I must be missing some important details. 

This is what AWS S3 documentation says about accessing buckets:
```
In a virtual-hosted–style URL, the bucket name is part of the domain name in the URL. For example:  

http://bucket.s3.aws-region.amazonaws.com.

http://bucket.s3.amazonaws.com

Note

Buckets created in Regions launched after January 30, 2019 are not reachable via the https://bucket.s3.amazonaws.com naming scheme.

In a virtual-hosted–style URL, you can use either of these endpoints. If you make a request to the http://bucket.s3.amazonaws.com endpoint, the DNS has sufficient information to route your request directly to the Region where your bucket resides.

For more information, see Virtual Hosting of Buckets.

 

In a path-style URL, the bucket name is not part of the domain. For example:

Region-specific endpoint, http://s3.aws-region.amazonaws.com/bucket

US East (N. Virginia) Region endpoint, http://s3.amazonaws.com/bucket

In a path-style URL, the endpoint you use must match the Region in which the bucket resides. For example, if your bucket is in the South America (São Paulo) Region, you must use the http://s3-sa-east-1.amazonaws.com/bucket endpoint. If your bucket is in the US East (N. Virginia) Region, you must use the http://s3.amazonaws.com/bucket endpoint.
```

I don't see anything in the docs that would prevent us from utilizing the existing `endpoint` field, and why we should introduce a new one.

Reply:
Given this impacts our ability to demo at Kubecon this week, and assuming this code works, would folks feel okay merging this as is and raising an issue to look into using `Endpoint` instead?

Reply:
According to my research, the bucketname is not necessary and we can set the `Endpoint` to the AWS region. I am about to push a commit to this branch with those changes.

https://docs.gitlab.com/ee/administration/job_artifacts.html#object-storage-settings
https://gitlab.com/charts/gitlab/blob/master/examples/objectstorage/rails.s3.yaml